tools/Datadog.Core.Tools/ConcurrencyHelper.cs

```diff
@@ -0,0 +1,152 @@
+                {
+                    Status = HelperStatus.Ready,
+                    Action = action,
+                    Name = friendlyName,
+                    Iterations = iterations,
+                });
+            }
+
+            Status = HelperStatus.Ready;
+        }
+
+        public async Task Start()
+        {
+            Status = HelperStatus.Running;
+
+            var registry = new ConcurrentQueue<Thread>();
+
+            Started = DateTime.Now;
+
+            var workers =
+                Levels
+                   .Select(level => new Thread(
+                               thread =>
+                               {
+                                   Interlocked.Increment(ref _remainingLevels);
+                                   _initiateLevelsEvent.Wait();
+                                   level.Started = DateTime.Now;
+                                   level.Status = HelperStatus.Running;
+
+                                   try
+                                   {
+                                       for (var i = 0; i < level.Iterations; i++)
+                                       {
+                                           try
+                                           {
+                                               level.Action();
+                                           }
+                                           catch (Exception ex)
+                                           {
+                                               level.Exceptions.Add(ex);
+                                           }
+                                       }
+                                   }
+                                   finally
+                                   {
+                                       level.Finished = DateTime.Now;
+                                       level.Status = HelperStatus.Finished;
+                                       Interlocked.Decrement(ref _remainingLevels);
+
+                                       if (_remainingLevels == 0)
+                                       {
+                                           // The run is finished
+                                           _testFinishedEvent.Set();
+                                       }
+                                   }
+                               }));
+
+            foreach (var worker in workers)
+            {
+                registry.Enqueue(worker);
+                worker.Start();
+            }
+
+            // Run everything
+            _initiateLevelsEvent.Set();
+
+            // Wait for everything to finish
+            _testFinishedEvent.Wait();
+
+            Finished = DateTime.Now;
+            Status = HelperStatus.Finished;
+
+            // Pause for the next run if this class is reused
+            _initiateLevelsEvent.Reset();
+
+            await Task.FromResult(0);
+        }
+
+        public Dictionary<string, int> GetExceptionSummary()
```

Comment:
Shouldn't this be a `ConcurrentDictionary` since it's populated from multiple threads?

Reply:
It's simplified for serialization purposes, I can do either probably

Reply:
The dictionary is created as a local variable and populated only within this method before it's returned. It's not populated from multiple threads unless I'm missing something and the dictionary is modified _after_ it's returned from this method. If it's not, maybe a `ReadOnlyDictionary` would express this intent better. Level of care: low.

Reply:
Since the method is called in a Task, from what I see, I think it would be in multiple threads. Easy enough to verify with TID logging

Reply:
It's definitely only called at the end. But I'll make it ReadOnly.